apps/builder/app/shared/commands-emitter.ts

```diff
@@ -0,0 +1,104 @@
+import { isHotkeyPressed } from "react-hotkeys-hook";
+import { atom, onMount } from "nanostores";
+import { $publisher, subscribe } from "~/shared/pubsub";
+
+type CommandMeta<CommandName extends string> = {
+  // @todo category, description
+  name: CommandName;
+  // default because hotkeys can be customized from ui
+  defaultHotkeys?: string[];
+};
+
+type CommandHandler = (source: string) => void;
+
+/**
+ * Command can be registered by builder, canvas or plugin
+ */
+type Command<CommandName extends string> = CommandMeta<CommandName> & {
+  /**
+   * Command handler accepting source where was triggered
+   * which is builder, canvas or plugin name
+   */
+  handler: CommandHandler;
+};
+
+/*
+ * expose command metas to synchronize between builder, canvas and plugins
+ */
+export const $commandMetas = atom(new Map<string, CommandMeta<string>>());
+
+export const createCommandsEmitter = <CommandName extends string>({
+  source,
+  commands,
+}: {
+  source: string;
+  // type only input to describe available commands from builder or other plugins
+  externalCommands?: CommandName[];
+  commands: Command<CommandName>[];
+}) => {
+  const commandMetas = new Map($commandMetas.get());
+  const commandHandlers = new Map<string, CommandHandler>();
+  for (const { handler, ...meta } of commands) {
+    commandMetas.set(meta.name, meta);
+    commandHandlers.set(meta.name, handler);
+  }
+
+  if (commands.length > 0) {
+    onMount($commandMetas, () => {
+      // use the next tick after subscription started
```

Comment:
Why use next tick? What if I start immediately

Reply:
Seems like the listener in sync-stores is executed after mount callback. So here scheduled to the end of execution so .set would log store as initialized. These all are hacks and need to figure out better way. Maybe with immerhin but without sending requests to server.

Reply:
Can you add this with example into comment, so everyone could get a problem. As the whole code looks strange

```js
// Read store
const commandMetas = new Map($commandMetas.get());

// on first subscriber read again?
onMount($commandMetas, () => {
    // use the next tick after subscription started
    Promise.resolve().then(() => {
      // @todo use patches to avoid race when both builder and canvas send commands
      $commandMetas.set(commandMetas);
    });
  });
```

Reply:
So onMount is needed for the 1st subscriber. And init for consequent.

Reply:
Yes, this is why we need to switch to patching instead of sending whole state. But in another PR. For now canvas just doesn't send empty commands